### PR TITLE
Warn when the token query param is used for auth

### PR DESCRIPTION
**Pull request by roncodingenthusiast**
### Description
- Emit a warning log message every time the token query param is used 
- Update docs to reflect the deprecation of the token query param 

Makes progress on #11917. Remaining work: fully remove in a future release.

#### Manual testing steps 
* Ran consul locally (make dev...)
* Tried a request with header
```
curl \
    --header "X-Consul-Token: <my token>" \
    "http://127.0.0.1:8500/v1/agent/members"

## No logs
```
* Tried a request with token without the header
```
curl \
    "http://127.0.0.1:8500/v1/agent/members?token=<my token>"

## log message
2023-01-18T15:37:36.989-0500 [WARN]  agent.http: This request is using a token query parameter; This is deprecated and will be removed in future Consul versions.: logUrl=/v1/agent/members?token=<hidden>
```
* Tried a request with token and the header present
```
curl \
    --header "X-Consul-Token: <my token>" \
    "http://127.0.0.1:8500/v1/agent/members?token=<my token>"

## log message
2023-01-18T15:48:18.764-0500 [WARN]  agent.http: This request is using a token query parameter; This is deprecated and will be removed in future Consul versions.: logUrl=/v1/agent/members?token=<hidden>
```

### Still todo
* [x] update tests that still use token query params to use header instead
* [x] Generate change-log

### PR Checklist

* [x] updated test coverage
* [x] external facing docs updated
* [x] not a security concern

### CODE REVIEW

File: website/content/api-docs/api-structure.mdx
Comment (jkirschner-hashicorp) [inline website/content/api-docs/api-structure.mdx:45]: Following up on @cthain 's comment from a previous PR...

Perhaps we should give explicit indication of when this will be removed. I don't expect all users to be familiar with our deprecation / removal policy.

Should we say something like:
```suggestion
Though the token could also be provided via a `?token=` query parameter,
that method is **highly discouraged** because the token can show up in access logs as part of the URL.
The `?token=` query parameter is deprecated and will be removed in Consul 1.17.
```

File: agent/http.go
Comment (jkirschner-hashicorp) [inline agent/http.go:371]: If we change the comment below, we might also say "will be removed in Consul 1.17" here.

File: website/content/docs/upgrading/upgrade-specific.mdx
Comment (jkirschner-hashicorp) [inline website/content/docs/upgrading/upgrade-specific.mdx:41]: This may imply that passing via the `token` query parameter was expected before Consul v1.15, which isn't accurate.

The `X-Consul-Token` header has been supported for a while and has been the recommended way. Now we're explicitly deprecating the token query param.

We should also mention the log message that indicates if the token query param is in use.

Perhaps change to something like:

````suggestion
Providing a Consul ACL token in API requests using the `token` query parameter is deprecated and will be removed in Consul 1.17.
Instead, the token should be provided using the `X-Consul-Token` header or with the Bearer scheme in the authorization header as described in the [API authentication documentation](/consul/api-docs/api-structure#authentication).

To check whether the `token` query parameter is in use, search your Consul agent logs for the message:
```shell-session
<INSERT WHAT THE MESSAGE IS>
```

Deprecated authentication using the `token` query parameter:

```shell-session
$ curl \
    http://127.0.0.1:8500/v1/agent/members?token=<consul token>
```

Recommended usage:

```shell-session
$ curl \
    --header "X-Consul-Token: <consul token>" \
    http://127.0.0.1:8500/v1/agent/members
```
````

File: .changelog/16009.txt
Comment (kisunji) [inline .changelog/16009.txt:2]: It's a little hard to discover but we have a set of "subsystems" we use in our changelogs. Adding something new won't break anything but it likely won't be grouped with similar PRs

```suggestion
acl: Warn when query token parameter is used for authentication.
```

File: agent/http.go
Comment (kisunji) [inline agent/http.go:370]: I don't foresee there ever being more than one `"token"` value but we should have this warn outside of the for loop since we wouldn't need to log more than once.

**Review by kisunji — APPROVED** [review]
:shipit:

File: website/content/api-docs/api-structure.mdx
Comment (im2nguyen) [inline website/content/api-docs/api-structure.mdx:45]:
```suggestion
~> **Security Note:** Though you could pass the token through a `?token=` query parameter, this method is highly discouraged because the token can show up in access logs as part of the URL. The `?token=` query parameter is deprecated and will be removed in Consul 1.17.
```

File: website/content/docs/upgrading/upgrade-specific.mdx
Comment (im2nguyen) [inline website/content/docs/upgrading/upgrade-specific.mdx:32]:
````suggestion
Instead, you should provide the token through the `X-Consul-Token` header or with the Bearer scheme in the authorization header as described in the [API authentication documentation](/consul/api-docs/api-structure#authentication).
	
Check whether you are using the `token` query parameter by searching your Consul agent logs for the message:

```shell-session hideClipboard
$ This request is using a token query parameter which is deprecated and will be removed in Consul 1.17
````
```

**Review by im2nguyen — APPROVED** [review]
a couple nits, but pre-approving so you're not blocked!

File: agent/http.go
Comment (jkirschner-hashicorp) [inline agent/http.go:371]: There's only one, so we should say "the" token query param. "a" suggests there might be multiple query parameters that could include a token, rather than that the query param is named literally "token". (At least by my read)

```suggestion
			httpLogger.Warn("This request is using the token query parameter "+
```

File: website/content/docs/upgrading/upgrade-specific.mdx
Comment (jkirschner-hashicorp) [inline website/content/docs/upgrading/upgrade-specific.mdx:31]:
```suggestion
$ This request is using the token query parameter which is deprecated and will be removed in Consul 1.17
```

File: website/content/docs/upgrading/upgrade-specific.mdx
Comment (jkirschner-hashicorp) [reply website/content/docs/upgrading/upgrade-specific.mdx:32]: "the token query param" rather than "a token query param"

File: .changelog/16009.txt
Comment (cthain) [inline .changelog/16009.txt:2]: What do you think about including the deprecation notice in the entry? I feel like it makes this more clear. Something like:

```suggestion
acl: Deprecate the `token` query parameter and warn when it is used for authentication.
```

File: agent/agent_endpoint_test.go
Comment (cthain) [inline agent/agent_endpoint_test.go:2027]: nit: The original request was using `httptest.NewRequest` instead of `http.NewRequest` as is done now. Just wondering if there is a reason to use the latter in this case instead?

File: website/content/docs/upgrading/upgrade-specific.mdx
Comment (cthain) [reply website/content/docs/upgrading/upgrade-specific.mdx:31]: nit: The suggestion in http.go says "This request **used** the token..." I think "used" is the word that we want here, not "uses".. Just pointing this out for consistency.

**Review by cthain — COMMENTED** [review]
Looking really good. I had a few minor comments/suggestions.

File: agent/agent_endpoint_test.go
Comment (roncodingenthusiast) [reply agent/agent_endpoint_test.go:2027]: no particular reason, I actually will reverse to `httptest` so I don't accidentally modify the behaviour. Good catch!

**Review by cthain — APPROVED** [review]
Great work on this PR Ronald :+1:

**Comment by jkirschner-hashicorp** [comment]
Awesome work :) Looks like engineering and education have both approved. Merge at will!

(Well, I guess the branch needs to be re-updated from main first / the CI pipeline has to finish first)

File: agent/http.go
Comment (jkirschner-hashicorp) [inline agent/http.go:373]: @kisunji , @cthain , @roncodingenthusiast : Should we strip the token from the `logUrl`?

**Review by jkirschner-hashicorp — CHANGES_REQUESTED** [review]
Placing a block on this until we discuss something I just left a comment on. (Still writing the comment.)

File: agent/http.go
Comment (jkirschner-hashicorp) [reply agent/http.go:373]: Does the token secret get stripped from the `logUrl` in the log output? Or does this log message actually include the token secret?

Comment (jkirschner-hashicorp) [reply agent/http.go:373]: Edit: I've been informed that the `logUrl` already obfuscates the token before the log is emitted

**Review by jkirschner-hashicorp — APPROVED** [review]
My previous review comment has been addressed 👍

**Comment by kisunji** [comment]
@roncodingenthusiast You'll have to OAuth into CircleCI with Github for the CI to run in this PR

**Pull request merged**
